### PR TITLE
Move Resource Barriers to D3D12 Binding

### PR DESCRIPTION
**Pull request by austinEng**
- Moves resource barriers from the D3D12 backend to the binding. Previously, resource transitions between `D3D12_RESOURCE_STATE_RENDER_TARGET` and `D3D12_RESOURCE_STATE_PRESENT` were done at the start and end of every render pass. This is incorrect and should only be done when the render target changes and buffers are swapped.

- With the binding handling the frame transitions, each frame has it's own `ID3D12CommandAllocator` so the next frame can be started before the current frame is done. This is what Microsoft shows to do in their frame buffering [example](https://github.com/Microsoft/DirectX-Graphics-Samples/blob/master/Samples/Desktop/D3D12HelloWorld/src/HelloFrameBuffering/D3D12HelloFrameBuffering.cpp)

- Fix which prevents a `Queue` from resetting an `ID3D12CommandAllocator` before its commands have completed on the GPU.

### CODE REVIEW

File: src/backend/d3d12/QueueD3D12.cpp [hunk 1]
```diff
@@ -32,9 +32,21 @@ namespace d3d12 {
...
         ));
         ASSERT_SUCCESS(commandList->Close());
+
+        ASSERT_SUCCESS(device->GetD3D12Device()->CreateFence(fenceValue, D3D12_FENCE_FLAG_NONE, IID_PPV_ARGS(&fence)));
+        fenceEvent = CreateEvent(nullptr, FALSE, FALSE, nullptr);
+        ASSERT(fenceEvent != nullptr);
     }
 
     void Queue::Submit(uint32_t numCommands, CommandBuffer* const * commands) {
+        const uint64_t currentFence = fenceValue++;
```
Comment (Kangz) [inline src/backend/d3d12/QueueD3D12.cpp]: nit: can you add a TODO saying that this stalls on the previous submit but we should fix it at some point?

File: examples/D3D12Binding.cpp [hunk 2]
```diff
@@ -152,14 +172,47 @@ class D3D12Binding : public BackendBinding {
         ComPtr<IDXGISwapChain3> swapChain;
         ComPtr<ID3D12DescriptorHeap> renderTargetViewHeap;
         ComPtr<ID3D12Resource> renderTargetResources[kFrameCount];
-        uint64_t fenceValues[kFrameCount] = { 0, 0 };
+        ComPtr<ID3D12CommandAllocator> commandAllocators[kFrameCount];
+        ComPtr<ID3D12GraphicsCommandList> commandList;
+        uint64_t fenceValues[kFrameCount] = {0};
```
Comment (Kangz) [inline examples/D3D12Binding.cpp]: Can you also address this comment from the other review? Especially since you introduced a framenumber.

> I found the code dealing with fenceValues hard to follow. To make it a bit simpler, you could have only one fence-updated variable, and call it something like "lastFrameFinishedSerial". You'll need another variable (not used in a fence) that's "currentFrameSerial". What do you think?

File: examples/D3D12Binding.cpp [hunk 3]
```diff
@@ -103,40 +103,60 @@ class D3D12Binding : public BackendBinding {
...
             fenceEvent = CreateEvent(nullptr, FALSE, FALSE, nullptr);
             ASSERT(fenceEvent != nullptr);
 
-            SetNextRenderTarget();
+            SetNextRenderTarget(frameIndex);
         }
         void SwapBuffers() override {
-            // Present the rendered frame
+            // Execute commands to present the current frame's render target
+            PresentFrame(frameIndex);
```
Comment (Kangz) [inline examples/D3D12Binding.cpp]: PresentFrame is only called here, either it could be inlined here, or the swapChain->Present call should be moved inside PresentFrame (and PresentFrame renamed to something like EnqueuePresentCommands).

File: examples/D3D12Binding.cpp [hunk 4]
```diff
@@ -152,14 +172,47 @@ class D3D12Binding : public BackendBinding {
...
+            commandList->ResourceBarrier(1, &resourceBarrier);
+
+            ASSERT_SUCCESS(commandList->Close());
+            ID3D12CommandList* commandLists[] = { commandList.Get() };
+            commandQueue->ExecuteCommandLists(_countof(commandLists), commandLists);
+            ASSERT_SUCCESS(commandList->Reset(commandAllocators[frameIndex].Get(), nullptr));
+        }
+
+        void PresentFrame(uint32_t frameIndex) {
+            D3D12_RESOURCE_BARRIER resourceBarrier;
```
Comment (Kangz) [inline examples/D3D12Binding.cpp]: I'm not sure what the semantics of D3D12 are, so this might not work. The code could be simplified by having SwapBuffers enqueue the following operations:

 - Transition from RENDER_TARGET to PRESENT
 - Present
 - Transition from PRESENT to RENDER_TARGET
 - Signal the fence.

Then SwapBuffers can be slightly simplified too to do the following:

 - Wait on the N -2 frame
 - Reset the allocators etc.
 - Enqueue work

File: src/backend/d3d12/QueueD3D12.cpp (hunk 1 above)
Comment (austinEng) [inline src/backend/d3d12/QueueD3D12.cpp]: Done

File: examples/D3D12Binding.cpp (hunk 2 above)
Comment (austinEng) [inline examples/D3D12Binding.cpp]: Updated with clearer variable names

File: examples/D3D12Binding.cpp (hunk 4 above)
Comment (austinEng) [inline examples/D3D12Binding.cpp]: Resource transitions could not be merged, but resetting allocators was simplified

File: examples/D3D12Binding.cpp (hunk 3 above)
Comment (austinEng) [inline examples/D3D12Binding.cpp]: inlined

**Comment by Kangz** [comment]
The commit changed indentation of most of D3D12Binding so the diff is really hard to see, can you look at it?

**Comment by austinEng** [comment]
> The commit changed indentation of most of D3D12Binding so the diff is really hard to see, can you look at it?

Fixed

File: examples/D3D12Binding.cpp [hunk 5]
```diff
@@ -103,65 +102,147 @@ class D3D12Binding : public BackendBinding {
...
-            ASSERT_SUCCESS(d3d12Device->CreateFence(fenceValues[frameIndex], D3D12_FENCE_FLAG_NONE, IID_PPV_ARGS(&fence)));
-            fenceValues[frameIndex]++;
+            // Get the initial render target and arbitrarily choose a "previous" render target that's different
+            previousRenderTargetIndex = renderTargetIndex = swapChain->GetCurrentBackBufferIndex();
+            previousRenderTargetIndex = renderTargetIndex == 0 ? 1 : 0;
+
+            // Initialize the current frame, the last completed frame, and all last frame each render target was used to 0
+            // so that it looks like we completed all previous frames
+            currentFrameNumber = 0;
+            for (uint32_t n = 0; n < kFrameCount; ++n) lastFrameRenderTargetWasUsed[n] = 0;
```
Comment (Kangz) [inline examples/D3D12Binding.cpp]: nit: braceless control flow is a bit dangerous.

File: examples/D3D12Binding.cpp [hunk 6]
```diff
@@ -103,65 +102,147 @@ class D3D12Binding : public BackendBinding {
...
+            {
+                ASSERT_SUCCESS(d3d12Device->CreateCommandList(
+                    0,
+                    D3D12_COMMAND_LIST_TYPE_DIRECT,
+                    commandAllocators[previousRenderTargetIndex].Get(),
+                    nullptr,
+                    IID_PPV_ARGS(&commandList)
+                ));
+
+                // Transition the first frame to be a render target
```
Comment (Kangz) [inline examples/D3D12Binding.cpp]: nit: you can put this comment before the block

File: examples/D3D12Binding.cpp [hunk 7]
```diff
@@ -103,65 +102,147 @@ class D3D12Binding : public BackendBinding {
...
+                ID3D12CommandList* commandLists[] = { commandList.Get() };
+                commandQueue->ExecuteCommandLists(_countof(commandLists), commandLists);
+            }
+
             ASSERT_SUCCESS(swapChain->Present(1, 0));
+            ASSERT_SUCCESS(commandQueue->Signal(fence.Get(), currentFrameNumber));
 
-            const uint64_t currentFenceValue = fenceValues[frameIndex];
-            ASSERT_SUCCESS(commandQueue->Signal(fence.Get(), currentFenceValue));
+            // Transition last frame's render target back to being a render target
```
Comment (Kangz) [inline examples/D3D12Binding.cpp]: This should be done before the signal, otherwise when the fence value is updated, the commands in commandAllocators[renderTargetIndex] might still be in flight.

**Comment by austinEng** [comment]
@Kangz Updated

File: examples/D3D12Binding.cpp [hunk 8]
```diff
@@ -103,65 +102,150 @@ class D3D12Binding : public BackendBinding {
...
+                resourceBarrier.Transition.StateAfter = D3D12_RESOURCE_STATE_RENDER_TARGET;
+                resourceBarrier.Transition.Subresource = D3D12_RESOURCE_BARRIER_ALL_SUBRESOURCES;
+                resourceBarrier.Type = D3D12_RESOURCE_BARRIER_TYPE_TRANSITION;
+                resourceBarrier.Flags = D3D12_RESOURCE_BARRIER_FLAG_NONE;
+                commandList->ResourceBarrier(1, &resourceBarrier);
+                ASSERT_SUCCESS(commandList->Close());
+                ID3D12CommandList* commandLists[] = { commandList.Get() };
+                commandQueue->ExecuteCommandLists(_countof(commandLists), commandLists);
+            }
+            ASSERT_SUCCESS(commandQueue->Signal(fence.Get(), currentFrameNumber));
```
Comment (Kangz) [inline examples/D3D12Binding.cpp]: The fence is signaled multiple times

**Comment by austinEng** [comment]
@Kangz This was rebased against `master`. Needs approval again

**Comment by Kangz** [comment]
Removed that restriction "branch must be up to date restriction", it seems too strong for now.
